### PR TITLE
cardano-cli transaction view: Add withdrawals

### PR DESCRIPTION
**Pull request by cblp**
1. Restructure transaction friendly-printer to use generic getTransactionBodyContent
2. Add support of all other fields in transaction friendly-printer:
   - withdrawals
   - metadata
   - (aux?) scripts

### CODE REVIEW

**Comment by cblp** [comment]
@Jimbo4350 I'm not sure loading scripts using V2 only is a good idea, but I don't know how to do it better. What do you think?

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 1]
```diff
@@ -11,156 +11,156 @@
...
+        ++  [ "certificates" .= friendlyCertificates txCertificates
+            | Just _ <- [certificatesSupportedInEra era]
+            ]
+        ++  [ "metadata" .= friendlyMetadata txMetadata
+            | Just _ <- [txMetadataSupportedInEra era]
+            ]
+        ++  [ "mint" .= friendlyMintValue txMintValue
+            | Right _ <- [multiAssetSupportedInEra era]
+            ]
+        ++  [ "update proposal" .= friendlyUpdateProposal txUpdateProposal
```
Comment (DavidEichmann) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:84]: Here and in other places we are making backwards incompatible changes to the output structure. Are we OK with this because it is just a "friendly" output?

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Please don't merge this PR until the Alonzo changes are in. Lots of things are being updated in cardano-api and this PR will make my current task more difficult if its merged.

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 2]
```diff
@@ -1261,29 +1489,53 @@ toShelleyWithdrawal withdrawals =
         [ (toShelleyStakeAddr stakeAddr, toShelleyLovelace value)
         | (stakeAddr, value) <- withdrawals ]
 
+fromShelleyWithdrawal :: Map (Shelley.RewardAcnt StandardCrypto) Shelley.Coin
+                      -> [(StakeAddress, Lovelace)]
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1503]: done

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 3]
```diff
@@ -1005,7 +1000,10 @@ instance Error (TxBodyError era) where
         | (k, err) <- errs ]
     displayError TxBodyMintAdaError =
       "Transaction cannot mint ada, only non-ada assets"
-
+    displayError TxBodyMintBeforeMaryError =
+      "Transaction can mint in Mary era or later"
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1007]: I think disallowing minting Ada and allowing minting are different things. Please correct me I miss something.

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 4]
```diff
@@ -1253,6 +1270,217 @@ makeShelleyTransactionBody era@ShelleyBasedEraMary
...
+      , txAuxScripts = TxAuxScriptsNone
+      , txWithdrawals
+      , txCertificates
+      , txUpdateProposal
+      , txMintValue = TxMintNone
+      }
+  where
+    adHash' =
+      maybeToStrictMaybe $ Ledger.hashAuxiliaryData @StandardShelley <$> auxData
+    txOuts = fromTxOut ShelleyBasedEraShelley <$> toList (Shelley._outputs body)
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/TxBody.hs:1304]: I extracted txIns for consistency with txOuts.

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 1 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:84]: Yes, this is just UI. API compatibility rules don't apply. I think, now the name is more correct.

**Comment by cblp** [comment]
@DavidEichmann are you OK with waiting for Alonzo?

**Comment by DavidEichmann** [comment]
@cblp That's fine with me. I think @ksaric might be waiting on these changes.

**Comment by ksaric** [comment]
> @cblp That's fine with me. I think @ksaric might be waiting on these changes.

Just need the `getTransactionBodyContent` for SMASH metadata, but will need that once the migration to `db-sync` is complete.
Will scream and shout if I reach that point and don't have it merged.

**Comment by Jimbo4350** [comment]
@cblp A way forward would be to only add changes necessary to implement and expose `getTransactionBodyContent` while stashing the other changes. After I've incorporated the Alonzo era changes we can come back and have a look at your other changes in this PR. cc: @ksaric

**Comment by cblp** [comment]
@Jimbo4350 "changes necessary to implement and expose getTransactionBodyContent" are almost all changes in this patch. The smaller part of the changes is about friendly printing and not so important.

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 5]
```diff
@@ -11,167 +11,203 @@
...
-import           Cardano.Api.Byron (TxBody (ByronTxBody))
-import           Cardano.Api.Shelley (TxBody (ShelleyTxBody), fromShelleyAddr)
-import           Cardano.Binary (Annotated)
-import qualified Cardano.Chain.UTxO as Byron
-import           Cardano.Ledger.Shelley as Ledger (ShelleyEra)
-import           Cardano.Ledger.ShelleyMA (MaryOrAllegra (Allegra, Mary), ShelleyMAEra)
-import qualified Cardano.Ledger.ShelleyMA.TxBody as ShelleyMA
-import           Ouroboros.Consensus.Shelley.Protocol.Crypto (StandardCrypto)
-import           Shelley.Spec.Ledger.API (Addr (..), TxOut (TxOut))
+                   AddressTypeInEra (ByronAddressInAnyEra, ShelleyAddressInEra),
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:21]: We should arrange the imports such that you can just have `import Cardano.Api` otherwise the imports will continue to get even more verbose.

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 6]
```diff
@@ -11,167 +11,203 @@
...
-    [ "inputs" .= inputs
-    , "outputs" .= fmap friendlyTxOutAllegra outputs
-    , "certificates" .= fmap textShow certificates
-    , "withdrawals" .= withdrawals
-    , "fee" .= txfee
-    , "validity interval" .= friendlyValidityInterval validity
-    , "update" .= fmap textShow update
-    , "auxiliary data hash" .= fmap textShow adHash
+friendlyTxBodyBS
+  :: IsCardanoEra era => Api.TxBody era -> Either Prelude.String ByteString
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:40]: Why the change from the original function that requires `CardanoEra era`?  We prefer value level evidence for the era. Likewise for the rest of the functions in this module.

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 1 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:81]: @cblp Please check with @disassembler and anyone else that may be using the pretty rendering to be sure.

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 7]
```diff
@@ -11,167 +11,203 @@
...
+            | Just _ <- [withdrawalsSupportedInEra era]
+            ]
+    Left err -> Left $ displayError err
+  where
+    era = cardanoEra @era
+
+friendlyValidityRange
+  :: forall era
+  .  IsCardanoEra era
+  => (TxValidityLowerBound era, TxValidityUpperBound era) -> [(Text, Value)]
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:95]: This can be make more generic using `validityLowerBoundSupportedInEra` and `validityUpperBoundSupportedInEra`

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 8]
```diff
@@ -11,167 +11,203 @@
...
+    era = cardanoEra @era
+    isLowerBoundSupported = isJust $ validityLowerBoundSupportedInEra era
+    isUpperBoundSupported = isJust $ validityUpperBoundSupportedInEra era
+
+friendlyWithdrawals :: TxWithdrawals ViewTx era -> Value
+friendlyWithdrawals TxWithdrawalsNone = Null
+friendlyWithdrawals (TxWithdrawals _ withdrawals) =
+  array
+    [ object
+        [ "address"     .= serialiseAddress addr
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:131]: :+1:

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 9]
```diff
@@ -11,167 +11,203 @@
...
-      [ "invalid before" .= invalidBefore
-      , "invalid hereafter" .= invalidHereafter
+friendlyTxOut :: TxOut era -> Value
+friendlyTxOut (TxOut addr amount) =
+  case addr of
+    AddressInEra ByronAddressInAnyEra _ ->
+      object $ ("address era" .= String "Byron") : common
+    AddressInEra (ShelleyAddressInEra _) (ShelleyAddress net cred stake) ->
+      object $
+        "address era"         .= String "Shelley"             :
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:146]: The usual pattern for lists is:
```
[  "address era" .= String "Shelley"
,  "network" .= net
...
]
```

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs [hunk 10]
```diff
@@ -921,13 +921,16 @@ runTxGetTxId txfile = do
 
 runTxView :: InputTxFile -> ExceptT ShelleyTxCmdError IO ()
 runTxView txfile = do
-  InAnyCardanoEra era txbody <-
+  InAnyCardanoEra _era txbody <-
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:924]: Why?

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Nice. Some comments below.

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 5 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:21]: Is implicit import list ok here?

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 6 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:40]: sorry, previously changed. fixed like everywhere

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 7 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs]: What do you mean?

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 9 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:130]: done

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs (hunk 10 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs]: reverted

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 5 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs]: Changed to implicit import

**Comment by cblp** [comment]
bors try

**Comment by cblp** [comment]
rebased after Alonzo, please review

File: cardano-cli/ChangeLog.md [hunk 11]
```diff
@@ -14,11 +14,11 @@
   - Auxiliary scripts (i.e. those included in the Tx auxiliary data, which are
     not required as transaction signers) must now be included with
     `--auxiliary-script-file` rather than with `--script-file`.
-  - Scripts witnessing txins, certificates, withdrawls and minting must now be
+  - Scripts witnessing txins, certificates, withdrawals and minting must now be
     paired with the thing they are witnessing. E.g.
     ```
     --certificate-file  $certfile --certificate-script-file $scriptfile
-    --tx-out $txout --minting-script-file $scriptfile
```
Comment (Jimbo4350) [inline cardano-cli/ChangeLog.md:21]: I'm not sure about the process for modifying the changelog cc: @LaurenceIO

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 12]
```diff
@@ -5,175 +5,190 @@
...
 import           Cardano.Prelude
 
-import           Data.Aeson (Object, Value (..), object, toJSON, (.=))
-import qualified Data.HashMap.Strict as HashMap
+import           Data.Aeson (Value (..), object, toJSON, (.=))
+import qualified Data.Aeson as Aeson
+import           Data.Yaml (array)
 import           Data.Yaml.Pretty (defConfig, encodePretty, setConfCompare)
 
-import           Cardano.Api as Api (AddressInEra (..),
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:19]: :100:

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 13]
```diff
@@ -5,175 +5,190 @@
...
+friendlyMetadataValue = \case
+  TxMetaNumber int   -> toJSON int
+  TxMetaBytes  bytes -> String $ textShow bytes
+  TxMetaList   lst   -> array $ map friendlyMetadataValue lst
+  TxMetaMap    m     ->
+    array
+      [array [friendlyMetadataValue k, friendlyMetadataValue v] | (k, v) <- m]
+  TxMetaText   text  -> toJSON text
+
+friendlyAuxScripts :: TxAuxScripts era -> Aeson.Value
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:188]: why not `friendlyAuxScripts :: CardanoEra era TxAuxScripts era -> Aeson.Value` and move `auxScriptsSupportedInEra` into this function? It will simplify `friendlyTxBody`.

**Review by Jimbo4350 — COMMENTED** [review]
Couple of comments but looking a lot better.

File: cardano-cli/ChangeLog.md (hunk 11 above)
Comment (cblp) [reply cardano-cli/ChangeLog.md:21]: I'm pretty sure this was a typo

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 13 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:188]: For declarativity, I wanted to have all keys in one place. `friendlyValidityRange` is the only exception because the keys depend on the era.

The style suggested by you may spread this logic into many small functions. I'm not sure this is good.

File: cardano-cli/ChangeLog.md (hunk 11 above)
Comment (cblp) [reply cardano-cli/ChangeLog.md:21]: I'm not sure now. Reverting.

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 14]
```diff
@@ -1573,10 +1571,6 @@ instance Error TxBodyError where
         | (k, err) <- errs ]
     displayError TxBodyMintAdaError =
       "Transaction cannot mint ada, only non-ada assets"
-    displayError TxBodyMintBeforeMaryError =
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:1576]: :+1:

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 15]
```diff
@@ -31,11 +31,11 @@ import qualified Shelley.Spec.Ledger.API as Shelley
 
 import           Cardano.CLI.Helpers (textShow)
 
-friendlyTxBodyBS :: CardanoEra era -> Api.TxBody era -> ByteString
+friendlyTxBodyBS :: CardanoEra era -> TxBody era -> ByteString
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:26]: This commit can be merged with the previous commit

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 16]
```diff
@@ -36,145 +28,139 @@ friendlyTxBodyBS era =
...
+        , "inputs"  .= friendlyInputs txIns
+        , "outputs" .= map friendlyTxOut txOuts
+        ]
+    ++  [ "certificates" .= friendlyCertificates txCertificates
+        | Just _ <- [certificatesSupportedInEra era]
+        ]
+    ++  [ "mint" .= friendlyMintValue txMintValue
+        | Right _ <- [multiAssetSupportedInEra era]
+        ]
+    ++  [ "update proposal" .= friendlyUpdateProposal txUpdateProposal
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:64]: https://github.com/input-output-hk/cardano-node/pull/2613#discussion_r701690475

@cblp There are already many small functions. What I suggest will remove all of the list comprehension noise.

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs [hunk 17]
```diff
@@ -162,5 +166,20 @@ friendlyTxOutValue = \case
   TxOutAdaOnly _ lovelace -> friendlyLovelace lovelace
   TxOutValue _ multiasset -> toJSON multiasset
 
+friendlyMetadata :: TxMetadataInEra era -> Aeson.Value
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Run/Friendly.hs:173]: Let's undo the alignment as per the style document.

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 13 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:188]: I'm not sure I understand what you mean by "have all keys in one place". You can leave the JSON keys in place. There are already several small functions, what I suggested will remove the list comprehension noise and the additional `++` operators.

File: cardano-cli/test/Test/Golden/TxView.hs [hunk 18]
```diff
@@ -16,11 +16,39 @@ txViewTests :: IO Bool
 txViewTests =
   checkSequential $
     Group "`transaction view` Goldens"
-      [ ("golden_view_shelley", golden_view_shelley)
+      [ ("golden_view_byron",   golden_view_byron)
+      , ("golden_view_shelley", golden_view_shelley)
       , ("golden_view_allegra", golden_view_allegra)
       , ("golden_view_mary",    golden_view_mary)
+      -- , ("golden_view_alonzo",  golden_view_alonzo)
```
Comment (Jimbo4350) [inline cardano-cli/test/Test/Golden/TxView.hs:23]: Is this not ready to be enabled?

**Review by Jimbo4350 — COMMENTED** [review]
Almost there

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 17 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:173]: just for `friendlyMetadata` function or for the whole file?

File: cardano-cli/test/Test/Golden/TxView.hs (hunk 18 above)
Comment (cblp) [reply cardano-cli/test/Test/Golden/TxView.hs:23]: I want to make it in a separate patch. This one is overloaded already.

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 13 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:188]: This is what I mean:
```
diff --git a/cardano-cli/src/Cardano/CLI/Run/Friendly.hs b/cardano-cli/src/Cardano/CLI/Run/Friendly.hs
index 347c89a4e..874afa3c5 100644
--- a/cardano-cli/src/Cardano/CLI/Run/Friendly.hs
+++ b/cardano-cli/src/Cardano/CLI/Run/Friendly.hs
@@ -48,9 +48,7 @@ friendlyTxBody
         , "fee"     .= friendlyFee txFee
         , "inputs"  .= friendlyInputs txIns
         , "outputs" .= map friendlyTxOut txOuts
-        ]
-    ++  [ "auxiliary scripts" .= friendlyAuxScripts txAuxScripts
-        | Just _ <- [auxScriptsSupportedInEra era]
+        , "auxiliary scripts" .= friendlyAuxScripts era txAuxScripts
         ]
     ++  [ "certificates" .= friendlyCertificates txCertificates
         | Just _ <- [certificatesSupportedInEra era]
@@ -185,8 +183,17 @@ friendlyMetadataValue = \case
       [array [friendlyMetadataValue k, friendlyMetadataValue v] | (k, v) <- m]
   TxMetaText   text  -> toJSON text
 
-friendlyAuxScripts :: TxAuxScripts era -> Aeson.Value
-friendlyAuxScripts = \case
+friendlyAuxScripts :: CardanoEra era -> TxAuxScripts era -> Aeson.Value
+friendlyAuxScripts era txAuxScripts =
+  case auxScriptsSupportedInEra era of
+    Nothing -> Aeson.Null
+    Just _ ->
+      case txAuxScripts of
+        TxAuxScriptsNone -> Null
+        TxAuxScripts _ scripts -> String $ textShow scripts
+
+friendlyAuxScripts' :: TxAuxScripts era -> Aeson.Value
+friendlyAuxScripts' = \case
   TxAuxScriptsNone       -> Null
   TxAuxScripts _ scripts -> String $ textShow scripts
 
```

Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:188]: My proposition gives more information, `null` has a meaning of "this field might be filled but is not", and an omitted field means "this cannot be filled".

> we should render `Null` in the case when it's not supported

I don't understand why should we render any value for a field when it has no sense. Why should we drop this information?

Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:188]: We can put all values in a single list, wrapping in `Unsupported | Supported a` datatype, and this filter out unsupported.

Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:188]: We should be explicit with the unsupported fields. We are explicit in other JSON renderings in the cli and report `Null` for those cases where we do not have information to render.

Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:191]: done

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 17 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:176]: done

Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:176]: There is still aligned code @cblp

**Review by Jimbo4350 — APPROVED** [review]
LGTM!

File: cardano-cli/src/Cardano/CLI/Run/Friendly.hs (hunk 17 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Run/Friendly.hs:176]: done now?

**Comment by cblp** [comment]
bors merge
